fil_fungible_token/src/token/mod.rs

```diff
@@ -278,8 +285,380 @@ where
+
+        // balances remained unchanged
+        let balance = token.balance_of(owner).unwrap();
+        assert_eq!(balance, TokenAmount::from(50));
+        let balance = token.balance_of(receiver).unwrap();
+        assert_eq!(balance, TokenAmount::zero());
+    }
+
+    #[test]
+    fn it_allows_delegated_transfer() {
+        let state_cid = Cid::default();
+        let mut token = Token::new(SharedMemoryBlockstore::new(), state_cid);
+        token.init_state().unwrap();
+
+        let owner = 1;
+        let receiver = 2;
+        let spender = 3;
+
+        // mint 100 for the owner
+        token.mint(owner, TokenAmount::from(100)).unwrap();
+        // approve 100 spending allowance for spender
+        token
+            .increase_allowance(owner, spender, TokenAmount::from(100))
+            .unwrap();
+        // spender makes transfer of 60 from owner -> receiver
+        token
+            .transfer(spender, owner, receiver, TokenAmount::from(60))
+            .unwrap();
+
+        // verify all balances are correct
+        let owner_balance = token.balance_of(owner).unwrap();
+        let receiver_balance = token.balance_of(receiver).unwrap();
+        let spender_balance = token.balance_of(spender).unwrap();
+        assert_eq!(owner_balance, TokenAmount::from(40));
+        assert_eq!(receiver_balance, TokenAmount::from(60));
+        assert_eq!(spender_balance, TokenAmount::zero());
+
+        // verify allowance is correct
+        let spender_allowance = token.allowance(owner, spender).unwrap();
+        assert_eq!(spender_allowance, TokenAmount::from(40));
+
+        // spender makes another transfer of 40 from owner -> self
+        token
+            .transfer(spender, owner, spender, TokenAmount::from(40))
+            .unwrap();
+
+        // verify all balances are correct
+        let owner_balance = token.balance_of(owner).unwrap();
+        let receiver_balance = token.balance_of(receiver).unwrap();
+        let spender_balance = token.balance_of(spender).unwrap();
+        assert_eq!(owner_balance, TokenAmount::zero());
+        assert_eq!(receiver_balance, TokenAmount::from(60));
+        assert_eq!(spender_balance, TokenAmount::from(40));
+
+        // verify allowance is correct
+        let spender_allowance = token.allowance(owner, spender).unwrap();
+        assert_eq!(spender_allowance, TokenAmount::zero());
+    }
+
+    #[test]
+    fn it_allows_revoking_allowances() {
+        let state_cid = Cid::default();
+        let mut token = Token::new(SharedMemoryBlockstore::new(), state_cid);
+        token.init_state().unwrap();
+
+        let owner = 1;
+        let receiver = 2;
+        let spender = 3;
+
+        // mint 100 for the owner
+        token.mint(owner, TokenAmount::from(100)).unwrap();
+        // approve 100 spending allowance for spender
+        token
+            .increase_allowance(owner, spender, TokenAmount::from(100))
+            .unwrap();
```

Comment:
Aside: I hate Rust's enforced line break heuristics. We can consider "fixing" it like https://github.com/filecoin-project/builtin-actors/blob/master/rustfmt.toml

Reply:
That looks good, I'll change the rules in a follow-up to minimise noise on this diff